provider/maas/environ.go

```diff
@@ -775,26 +775,89 @@ func (env *maasEnviron) spaceNamesToSpaceInfo(
 	return positiveSpaceIds, negativeSpaceIds, nil
 }
 
+// networkSpaceRequirements combines the space requirements for the application
+// bindings and the specified constraints and returns back a set of provider
+// space IDs for which a NIC needs to be provisioned in the instance we are
+// about to launch and a second (negative) set of space IDs that must not be
+// present in the launched instance NICs.
+func (env *maasEnviron) networkSpaceRequirements(ctx context.ProviderCallContext, endpointToProviderSpaceID map[string]corenetwork.Id, cons constraints.Value) (set.Strings, set.Strings, error) {
+	positiveSpaceIds := set.NewStrings()
+	negativeSpaceIds := set.NewStrings()
+
+	// Iterate the application bindings and add each bound space ID to the
+	// positive space set.
+	for epName, providerSpaceID := range endpointToProviderSpaceID {
+		// The default space is applied at deploy time to all endpoints
+		// that are not explicitly bound but should be ignored when
+		// collecting the spaces we need. Consider this edge case:
+		// an explicit space binding is provided for all endpoints as
+		// well as a default one that is *different* from all others.
+		// In this case, we shouldn't request for a NIC in the default
+		// space as it is not actually going to be used.
+		if epName == "" {
```

Comment:
I have added this extra guard to prevent the creation of an unused interface in the edge case where all endpoints are explicitly bound to a space different than the default value.